pallets/gear-messenger/src/migrations.rs

```diff
@@ -0,0 +1,327 @@
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Config, DispatchStash, Dispatches, Pallet, Waitlist};
+use common::storage::{Interval, LinkedNode};
+use frame_support::{
+    traits::{Get, GetStorageVersion, OnRuntimeUpgrade},
+    weights::Weight,
+};
+use gear_core::{
+    ids::MessageId,
+    message::{ContextStore, StoredDispatch},
+};
+use sp_std::marker::PhantomData;
+#[cfg(feature = "try-runtime")]
+use {
+    frame_support::codec::{Decode, Encode},
+    sp_std::vec::Vec,
+};
+
+pub struct MigrateToV3<T: Config>(PhantomData<T>);
+
+impl<T: Config> MigrateToV3<T> {
+    fn migrate_context_store(ctx: v2::ContextStore) -> ContextStore {
+        ContextStore {
+            outgoing: ctx.outgoing,
+            reply: ctx.reply,
+            initialized: ctx.initialized,
+            reservation_nonce: ctx.reservation_nonce,
+            system_reservation: ctx.system_reservation,
+        }
+    }
+}
+
+impl<T: Config> OnRuntimeUpgrade for MigrateToV3<T> {
+    fn on_runtime_upgrade() -> Weight {
+        let current = Pallet::<T>::current_storage_version();
+        let onchain = Pallet::<T>::on_chain_storage_version();
+
+        log::info!(
+            "🚚 Running migration with current storage version {current:?} / onchain {onchain:?}"
+        );
+
+        // 1 read for on chain storage version
+        let mut weight = T::DbWeight::get().reads(1);
+
+        if current == 3 && onchain == 2 {
+            Waitlist::<T>::translate(
+                |_, _, store: (v2::StoredDispatch, Interval<T::BlockNumber>)| {
+                    weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+                    Some((
+                        StoredDispatch {
+                            kind: store.0.kind,
+                            message: store.0.message,
+                            context: store.0.context.map(Self::migrate_context_store),
+                        },
+                        store.1,
+                    ))
+                },
+            );
+
+            Dispatches::<T>::translate(|_, store: LinkedNode<MessageId, v2::StoredDispatch>| {
+                weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+                Some(LinkedNode {
+                    next: store.next,
+                    value: StoredDispatch {
+                        kind: store.value.kind,
+                        message: store.value.message,
+                        context: store.value.context.map(Self::migrate_context_store),
+                    },
+                })
+            });
+
+            DispatchStash::<T>::translate(
```

Comment:
btw you don't need to actually migrate dispatch stash because context there is always None. In this migrations you can change its data structure into sth like StoredDelayedDispatch { kind, msg }
In that case assert in tests that they're always empty and propagate here log::error in decode err case